### PR TITLE
add support for aws_lambda_source_event_mapping

### PR DESCRIPTION
**Pull request by moadibfr**
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | yes
| ⚠ Deprecations?   | no
| ❌ BC Break       | no
| 🔗 Related issues | #250 
| ❓ Documentation  | yes

## Description

add support for aws_lambda_source_event_mapping

### CODE REVIEW

File: pkg/iac/deserializers.go [hunk 1]
```diff
@@ -58,8 +58,8 @@ func Deserializers() []deserializer.CTYDeserializer {
 		awsdeserializer.NewCloudfrontDistributionDeserializer(),
 		awsdeserializer.NewKMSKeyDeserializer(),
 		awsdeserializer.NewKMSAliasDeserializer(),
-
 		ghdeserializer.NewGithubRepositoryDeserializer(),
 		ghdeserializer.NewGithubTeamDeserializer(),
+		awsdeserializer.NewLambdaEventSourceMappingDeserializer(),
```
Comment (wbeuil) [inline pkg/iac/deserializers.go:63]: I would write it just below other aws deserializers

**Review by wbeuil — DISMISSED** [review]
Just one small typo

File: pkg/remote/aws/lambda_event_source_mapping_supplier.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	}
+}
+
+func (s LambdaEventSourceMappingSupplier) Resources() ([]resource.Resource, error) {
+	functions, err := s.client.ListAllLambdaEventSourceMappings()
+	if err != nil {
+		return nil, remoteerror.NewResourceEnumerationError(err, resourceaws.AwsLambdaEventSourceMappingResourceType)
+	}
+	results := make([]cty.Value, 0)
+	if len(functions) > 0 {
```
Comment (eliecharra) [inline pkg/remote/aws/lambda_event_source_mapping_supplier.go:41]: useless if there

File: pkg/remote/aws/lambda_event_source_mapping_supplier.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+		}
+	}
+	return s.deserializer.Deserialize(results)
+}
+
+func (s LambdaEventSourceMappingSupplier) readLambdaEventSourceMapping(sourceMappingConfig lambda.EventSourceMappingConfiguration) (cty.Value, error) {
+	resFunction, err := s.reader.ReadResource(
+		terraform.ReadResourceArgs{
+			Ty: resourceaws.AwsLambdaEventSourceMappingResourceType,
+			ID: aws.StringValue(sourceMappingConfig.UUID),
```
Comment (eliecharra) [inline pkg/remote/aws/lambda_event_source_mapping_supplier.go:60]: UUID can really be nil there ? I don't think `aws.StringValue` is very usefull there

File: pkg/remote/aws/lambda_event_source_mapping_supplier.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
+package aws
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/cloudskiff/driftctl/pkg/remote/aws/repository"
+	remoteerror "github.com/cloudskiff/driftctl/pkg/remote/error"
+
```
Comment (eliecharra) [inline pkg/remote/aws/lambda_event_source_mapping_supplier.go:7]: There is spaces between your imports (not only in this file)

File: pkg/resource/aws/aws_lambda_event_source_mapping.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	LastModified                   *string `cty:"last_modified" computed:"true" diff:"-"`
+	LastProcessingResult           *string `cty:"last_processing_result" computed:"true" diff:"-"`
+	MaximumBatchingWindowInSeconds *int    `cty:"maximum_batching_window_in_seconds"`
+	MaximumRecordAgeInSeconds      *int    `cty:"maximum_record_age_in_seconds" computed:"true"`
+	MaximumRetryAttempts           *int    `cty:"maximum_retry_attempts" computed:"true"`
+	ParallelizationFactor          *int    `cty:"parallelization_factor" computed:"true"`
+	StartingPosition               *string `cty:"starting_position" diff:"-"`
+	StartingPositionTimestamp      *string `cty:"starting_position_timestamp" diff:"-"`
+	State                          *string `cty:"state" computed:"true" diff:"-"`
+	StateTransitionReason          *string `cty:"state_transition_reason" computed:"true"`
```
Comment (eliecharra) [inline pkg/resource/aws/aws_lambda_event_source_mapping.go:23]: I think that we could ignore this field too if we ignore the state ?

**Review by eliecharra — CHANGES_REQUESTED** [review]

File: pkg/remote/aws/lambda_event_source_mapping_supplier.go (hunk 3 above)
Comment (moadibfr) [reply pkg/remote/aws/lambda_event_source_mapping_supplier.go:60]: I shouldn't but we have seen strange errors so I think it's a good idea to not crash ?

File: pkg/remote/aws/lambda_event_source_mapping_supplier.go (hunk 4 above)
Comment (moadibfr) [reply pkg/remote/aws/lambda_event_source_mapping_supplier.go:7]: yeah it's go import way or organizing import ?

File: pkg/resource/aws/aws_lambda_event_source_mapping.go (hunk 5 above)
Comment (moadibfr) [reply pkg/resource/aws/aws_lambda_event_source_mapping.go:23]: yeah you're right I forgot to add it there

File: pkg/remote/aws/lambda_event_source_mapping_supplier.go (hunk 3 above)
Comment (eliecharra) [reply pkg/remote/aws/lambda_event_source_mapping_supplier.go:60]: If we want to handle this strange case we should not call `ReadResource` with a nil value IMHO, we should handle this case explicitly before in the loop, and maybe send a `log.Warning` as something went wrong

File: pkg/iac/deserializers.go (hunk 1 above)
Comment (moadibfr) [reply pkg/iac/deserializers.go:63]: ok but the list is already unordered ^^

File: pkg/remote/aws/lambda_event_source_mapping_supplier.go (hunk 4 above)
Comment (moadibfr) [reply pkg/remote/aws/lambda_event_source_mapping_supplier.go:7]: Hmm maybe it's goland I don't really know I removed them

Comment (eliecharra) [reply pkg/remote/aws/lambda_event_source_mapping_supplier.go:7]: I don't know what is the standard way to organize imports, but we should follow it and especially we should all use the same logic.

By default it seems that Goland split stdlib pkg (like `fmt`) from others and is using gofmt.

image

When using `go fmt -local` it groups imports by pkg and add whitespaces between them.

I also have goimport as file watcher in goland and the same gofmt logic seems to be applied.

I don't really care which option we take but we should avoid unnecessary conflicts in imports.

Comment (moadibfr) [reply pkg/remote/aws/lambda_event_source_mapping_supplier.go:7]: yeah I didn't see that go fmt had an other way of organizing, I removed the newline but I have no preference on what to use

**Review by eliecharra — APPROVED** [review]

**Pull request merged**